cartography/intel/rapid7/util.py

```diff
@@ -0,0 +1,445 @@
+
+    logger.debug("extract_rapid7_configurations_resourcegroup: %s", resource_group)
+    return resource_group
+
+
+def extract_rapid7_configurations_resourceid(configurations: Any[str, List]) -> str:
+    """
+    Extract from json, specific keyvalue pair: azure resource id
+
+    Input: dataframe row configurations column
+    """
+
+    if configurations is None or not isinstance(configurations, list):
+        return ""
+
+    resource_id = ""
+    for line in configurations:
+        if line["name"] == "azure":
+            logger.debug("line: %s", line)
+            if isinstance(line["value"], str):
+                azure = json.loads(line["value"])
+            else:
+                azure = line["value"]
+            resource_id = azure["resourceId"]
+
+    logger.debug("extract_rapid7_configurations_resourceid: %s", resource_id)
+    return resource_id
+
+
+def extract_rapid7_configurations_instanceid(configurations: Any[str, List]) -> str:
+    """
+    Extract from json, specific keyvalue pair: azure instance id
+
+    Input: dataframe row configurations column
+    """
+
+    if configurations is None or not isinstance(configurations, list):
+        return ""
+
+    instance_id = ""
+    for line in configurations:
+        if line["name"] == "azure":
+            logger.debug("line: %s", line)
+            if isinstance(line["value"], str):
+                azure = json.loads(line["value"])
+            else:
+                azure = line["value"]
+            instance_id = azure["instanceId"]
+
+    logger.debug("extract_rapid7_configurations_instanceid: %s", instance_id)
+    return instance_id
+
+
+def rapid7_hosts_from_getassets(
+    authorization: Tuple[str, str, str, str, str, int],
+    limit: int = 6000,
+    page: int = 0,
+    # examples: "id,ASC", "id,DESC"
+    sort: str = "",
+    nexpose_timeout: int = 60,
+) -> List:
+    """
+    Get Nexpose inventory full data to array
+    https://help.rapid7.com/insightvm/en-us/api/index.html#operation/getAssets
+
+    To filter, use https://help.rapid7.com/insightvm/en-us/api/index.html#operation/findAssets
+
+    Warning! if very large assets count (>10k, likely depending on assets details), it
+    will have impact on memory needed for processing and could trigger linux oom-killer.
+    Possible option: have the server generate a report and just retrieve it.
+    https://help.rapid7.com/insightvm/en-us/api/index.html#operation/downloadReport
+    """
+
+    nexpose_verify_cert = authorization[3]
+    logger.info("nexpose_verify_cert: %s", nexpose_verify_cert)
+    if nexpose_verify_cert.lower() != "true":
+        nexpose_verify_cert2 = False
+        logger.warning("Requested to not verify certificate (%s)", nexpose_verify_cert)
+    else:
+        nexpose_verify_cert2 = True
+
+    count = total_resources = 0
+    size_interval = 500
+    result_array: List[Any] = []
+    df_rapid7_tmp = pandas.DataFrame()
+    while count < limit:
```

Comment:
Why do we have a limit? Is it because of API ratelimitting? If that's the case, consider using the `@backoff` decorator.
https://github.com/lyft/cartography/blob/master/cartography/util.py#L206

Reply:
No API rate-limiting AFAIK, more response body size and experience that it is more likely to fail if more depending on context. at least, not on doc https://help.rapid7.com/insightvm/en-us/api/index.html#operation/findAssets

Another reason to use the report option.